### ISSUE TITLE
Server Side Batching

### ISSUE THREAD

**Issue by ericmcbride** [hyperium/tonic#276]
Im familiar with GRPC Interceptors and have looked at Tower Layers, but im wondering what would be the best way to do server side batching on incoming requests?  Ive been reading through the source code, and the examples (which are awesome) but I was wondering if theres any way I could get a handle to incoming requests and make a buffer/vector with them the submit them to a corresponding GRPC endpoint

**Comment by LucioFranco** [comment]
@ericmcbride could you explain more what you mean by buffer the incoming requests? In theory you can always place a tower layer in-between hyper and tonic (the transport stuff is optional). Though with h2 being a multiplexed and streaming protocol this doesn't seem like there is terribly much advantage.

**Comment by ericmcbride** [comment]
So I have a use case (using a lib for GPU processing for Machine Learning) where I need to collect incoming requests into a vector then post to a grpc endpoint, if a certain flag is passed.  I saw inside of interop where you can wrap a service in another service, so I was thinking maybe collecting the requests there, and posting to that endpoint.  That endpoint would have a call to a function, that would do a GPU lookup for a modlel.  

If you post a batch of requests to the GPU, you can get a large batch of responses back in the same time itd take a singular request. Sorry Im vauge in details but 

TL;DR
I need to collect incoming requests into a vector, then post to a GRPC endpoint.

**Comment by LucioFranco** [comment]
@ericmcbride that totally makes sense! What is the type of the inbound requests is it a simple unary grpc request that you then want to batch into another endpoints grpc unary request to send it as a vec.

many inbound -> single outbound?

**Comment by ericmcbride** [comment]
its a simple unary request.  

So itd be like fan in (to make a vector of whatever I need), make the request(GPU Batch Call) and send a response back.  In the future itd be cool at the end to fan back out to the respected singular requests.  I don't have enough insight just yet on hyper / tower / tonic, but im having alot of fun digging through.

**Comment by LucioFranco** [comment]
@ericmcbride I would probably not try to use tower to do this but I would most likely just take advantage of async/await. So you could for example use a mpsc channel to fan-in requests to a background task. The background task collections requests off of the channel and can then batch them and send them on the outbound rpc. The real key here is when you send the original request on the mpsc channel also create a oneshot channel. This will allow the background task to send a response back to that specific request. Using this with async rust should be very efficient and should work quite well.

Here is some code (tonic is out dated here) but should show the same pattern. Hopefully that helps!

https://github.com/LucioFranco/kv/blob/master/src/network.rs#L252

**Comment by ericmcbride** [comment]
Cool thanks for the link!  I was having issues trying to grab the inbound requests (i was trying to use tower::service_fn and tx.send within the closure).  how would you suggest grabbing the inbound request in context of the server running.  After i figure that out, you have done more then enough! Thanks again!

**Comment by LucioFranco** [comment]
Is there a reason you want to grab the inbound request directly in the transport instead of using the tonic codgen? To me nothing in this tells me that you should use tower directly (unless if that is your goal that is totally fine!). I think it makes more sense to avoid tower and write it directly with async/await.

**Comment by ericmcbride** [comment]
Nah i just dont have any direction TBH at this point, i thought the only way you could grab an inbound request was via transport.

**Comment by LucioFranco** [comment]
You can! though basically all that tonic does is map raw http requests to grpc ones and then you are within a future that allows you to async respond. Which is what tower etc would give you as well but its much more lower level and is really designed for super generic things like retries and timeouts. Otherwise, it makes sense to write this logic directly in the handler that tonic generates.

**Comment by ericmcbride** [comment]
AHHHHH! Lol this is initially where i started at and have went full circle! I was like should I just batch them in the actual handler, then something about that felt weird, should've went with my gut instinct.  Im gonna close out this issue.  Thanks a ton for your gudiance!

**Comment by LucioFranco** [comment]
No problem! Good luck :)
